### DISCUSSION TITLE
How to tell Dagster that some partitions are already filled in

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#12561] [Q&A] [answered]
Is there a way to tell Dagster that the first thousand partitions of a time-partitioned asset are already completed ? If I copy data manually into a table?

I'd like to mark these assets and partitions as materialized without actually running the code that overwrites their contents.

**Comment by mjclarke94** [comment]
Is there a way of copying over the state of the postgresdb for branch deployments? We're following the guide [here](https://docs.dagster.io/guides/dagster/branch_deployments) to clone our entire production database for each branch deployment, but we'd obviously rather not need to then create dummy materialisations for every single partition of every asset when we do so.

**Reply by sryza** [reply]
I don't believe this is currently possible, but the request makes sense. I forwarded this to our branch deployment folks.

**Comment by sryza** [comment]
### Old, deprecated answer



The best way to do this is to write a job that directly generates `AssetMaterialization`s to fill in the missing partitions.

```python
from dagster import asset, AssetMaterialization, DailyPartitionsDefinition, op, job, Definitions

daily_partitions_def = DailyPartitionsDefinition(start_date="2020-01-01")


@asset(partitions_def=daily_partitions_def)
def my_venerable_asset():
    ...


@op(config_schema={"start_date": str, "end_date": str})
def record_materializations(context):
    all_partition_keys = daily_partitions_def.get_partition_keys()

    start_index = all_partition_keys.index(context.op_config["start_date"])
    end_index = all_partition_keys.index(context.op_config["end_date"], start_index)

    for partition_key in all_partition_keys[start_index : end_index + 1]:
        context.log_event(
            AssetMaterialization(asset_key=my_venerable_asset.key, partition=partition_key)
        )


@job
def record_materializations_job():
    record_materializations()


defs = Definitions(
    assets=[my_venerable_asset],
    jobs=[record_materializations_job],
)
```

Running the job over a range of partitions will create materialization events for each partition, which lets Dagster understand that they're already materialized.

**Reply by sryza** [reply]
Here's an issue for making this a more first-class capability: https://github.com/dagster-io/dagster/issues/8795

**Comment by sryza** [comment, marked as answer]
This is now a native capability of Dagster: https://github.com/dagster-io/dagster/discussions/17847#discussioncomment-7523864
